### PR TITLE
feat: allow lazy binary download

### PR DESCRIPTION
**Pull request by yesnault**
if there is no binary in docker image, there will be auto downloaded
from github or artifactory if configured.

Signed-off-by: Yvonnick Esnault <yvonnick.esnault@corp.ovh.com>

1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16373.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

File: engine/api/download/download.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		if err := archiver.DefaultTarGz.Unarchive(path.Join(conf.Directory, filenameToDownload), conf.Directory); err != nil {
+			return sdk.WrapError(err, "unarchive %s failed", filenameToDownload)
+		}
+	}
+
+	return nil
+}
+
+func GetBinaryFromArtifactory(conf Conf, filename string) error {
+	jlog.SetLogger(jlog.NewLogger(jlog.ERROR, os.Stdout))
```
Comment (sguiheux) [inline engine/api/download/download.go:101]: Maybe logger must be set during jfrog client creation

File: engine/api/download/download.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+	artiClient, err := art.CreateArtifactoryClient(conf.ArtifactoryURL, conf.ArtifactoryToken)
```
Comment (sguiheux) [inline engine/api/download/download.go:102]: Do not use contrib package in API.
Maybe refact/reuse code in package api/integration/artifact_manager package

**Review by sguiheux — CHANGES_REQUESTED** [review]

File: engine/api/api.go [hunk 3]
```diff
@@ -76,9 +77,17 @@ type Configuration struct {
...
 	} `toml:"cache" comment:"######################\n CDS Cache Settings \n#####################" json:"cache"`
-	Directories struct {
-		Download string `toml:"download" default:"/var/lib/cds-engine" json:"download"`
-	} `toml:"directories" json:"directories"`
+	Download struct {
+		Directory          string `toml:"directory" default:"/var/lib/cds-engine" json:"directory" comment:"this directory contains cds binaries. If it's empty, cds will download binaries from GitHub (property downloadFromGitHub) or from an artifactory instance (property artifactory) to it"`
+		SupportedOSArch    string `toml:"supportedOSArch" default:"" json:"supportedOSArch" commented:"true" comment="example: darwin/amd64,darwin/arm64,linux/amd64,windows/amd64"`
+		DownloadFromGitHub bool   `toml:"downloadFromGitHub" default:"true" json:"downloadFromGitHub" comment:"allow downloading binaries from GitHub"`
+		Artifactory        struct {
+			URL        string `toml:"url" default:"https://your-artifactory" json:"url" comment:"URL of your artifactory"`
```
Comment (fsamin) [inline engine/api/api.go:85]:
```suggestion
			URL        string `toml:"url" default:"https://your-artifactory/artifactory" json:"url" comment:"URL of your artifactory"`
```

File: engine/api/download/download.go (hunk 1 above)
Comment (fsamin) [reply engine/api/download/download.go:101]: you should redirect the log to our logger. 
probably have to implement your own jlog.Logger to wrap calls to our logger

File: engine/api/download/download.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	if err != nil {
+		return fmt.Errorf("unable to create artifactory client: %v", err)
```
Comment (fsamin) [inline engine/api/download/download.go:104]: use errors.New() instead
from pkg/errors

File: engine/api/download/download.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+	}
+
+	params := services.NewDownloadParams()
+	params.Pattern = fmt.Sprintf("%s/%s/%s/%s", conf.ArtifactoryRepository, conf.ArtifactoryPath, sdk.CDSVERSION, filename)
+	params.Target = strings.TrimSuffix(conf.Directory, "/") + "/"
```
Comment (fsamin) [inline engine/api/download/download.go:109]: weird

File: engine/ui/ui.go [hunk 6]
```diff
@@ -336,24 +336,20 @@ func (s *Service) downloadStaticFilesFromGitHub(ctx context.Context, version str
 	if _, err := os.Stat(s.Cfg.Staticdir); os.IsNotExist(err) {
 		log.Info(ctx, "ui> creating directory %s", s.Cfg.Staticdir)
 		if err := os.Mkdir(s.Cfg.Staticdir, 0740); err != nil {
-			return fmt.Errorf("Error while creating directory: %v", err)
+			return fmt.Errorf("error while creating directory: %v", err)
```
Comment (fsamin) [inline engine/ui/ui.go:339]: errors.New from pkg/errors

File: engine/ui/ui.go [hunk 7]
```diff
@@ -336,24 +336,20 @@ func (s *Service) downloadStaticFilesFromGitHub(ctx context.Context, version str
...
-	urlFiles := fmt.Sprintf("https://github.com/ovh/cds/releases/download/%s/ui.tar.gz", version)
-	if version == "latest" {
-		var err error
-		urlFiles, err = s.Client.DownloadURLFromGithub("ui.tar.gz")
-		if err != nil {
-			return fmt.Errorf("Error while getting ui.tar.gz from Github err:%s", err)
-		}
+	urlFiles, err := sdk.DownloadURLFromGithub("ui.tar.gz", version)
+	if err != nil {
+		return fmt.Errorf("error while getting ui.tar.gz from GitHub err:%s", err)
```
Comment (fsamin) [inline engine/ui/ui.go:345]: errors.New from pkg/errors

File: engine/ui/ui.go [hunk 8]
```diff
@@ -336,24 +336,20 @@ func (s *Service) downloadStaticFilesFromGitHub(ctx context.Context, version str
...
 	}
 
 	log.Info(ctx, "ui> Downloading from %s...", urlFiles)
 
 	resp, err := http.Get(urlFiles)
 	if err != nil {
-		return fmt.Errorf("Error while getting ui.tar.gz from GitHub: %v", err)
+		return fmt.Errorf("error while getting ui.tar.gz from GitHub: %v", err)
```
Comment (fsamin) [inline engine/ui/ui.go:352]: errors.New from pkg/errors

File: engine/ui/ui.go [hunk 9]
```diff
@@ -362,7 +358,7 @@ func (s *Service) downloadStaticFilesFromGitHub(ctx context.Context, version str
 	}
 
 	if resp.StatusCode != 200 {
-		return fmt.Errorf("Error while getting ui.tar.gz from GitHub - HTTP code: %d", resp.StatusCode)
+		return fmt.Errorf("error while getting ui.tar.gz from GitHub - HTTP code: %d", resp.StatusCode)
```
Comment (fsamin) [inline engine/ui/ui.go:361]: errors.New from pkg/errors

File: engine/api/api.go [hunk 10]
```diff
@@ -76,9 +77,17 @@ type Configuration struct {
 			Password string `toml:"password" json:"-"`
 		} `toml:"redis" comment:"Connect CDS to a redis cache If you more than one CDS instance and to avoid losing data at startup" json:"redis"`
 	} `toml:"cache" comment:"######################\n CDS Cache Settings \n#####################" json:"cache"`
-	Directories struct {
-		Download string `toml:"download" default:"/var/lib/cds-engine" json:"download"`
-	} `toml:"directories" json:"directories"`
+	Download struct {
+		Directory          string `toml:"directory" default:"/var/lib/cds-engine" json:"directory" comment:"this directory contains cds binaries. If it's empty, cds will download binaries from GitHub (property downloadFromGitHub) or from an artifactory instance (property artifactory) to it"`
+		SupportedOSArch    string `toml:"supportedOSArch" default:"" json:"supportedOSArch" commented:"true" comment="example: darwin/amd64,darwin/arm64,linux/amd64,windows/amd64"`
```
Comment (fsamin) [inline engine/api/api.go:82]: why it is not an array ?

**Review by richardlt — APPROVED** [review]

File: sdk/download.go [hunk 11]
```diff
@@ -125,3 +173,142 @@ func GetContentType(resp *http.Response) string {
...
+			if *asset.Name == filename {
+				return *asset.BrowserDownloadURL, nil
+			}
+		}
+	}
+
+	text := fmt.Sprintf("Invalid Artifacts on latest release (%s). Please try again in few minutes.\n", filename)
+	text += fmt.Sprintf("If the problem persists, please open an issue on %s\n", URLGithubIssues)
+	text += fmt.Sprintf("You can manually download binary from latest release: %s\n", URLGithubReleases)
+	return "", fmt.Errorf(text)
```
Comment (fsamin) [inline sdk/download.go:245]: errors.New from pkg/errors

File: sdk/version.go [hunk 12]
```diff
@@ -9,6 +9,9 @@ var (
 	//VERSION is set with -ldflags "-X github.com/ovh/cds/sdk.VERSION=$(VERSION)"
 	VERSION = "snapshot"
 
+	//CDSVERSION is set with -ldflags "-X github.com/ovh/cds/sdk.CDSVERSION=$(CDSVERSION)"
+	CDSVERSION = "snapshot"
```
Comment (fsamin) [inline sdk/version.go:13]: keep only VERSION const

**Review by fsamin — CHANGES_REQUESTED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16380.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16381.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16384.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16385.0 ✘  
* Build
  * Build and Package All ✘  
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

File: engine/api/download/download.go (hunk 5 above)
Comment (yesnault) [reply engine/api/download/download.go:141]: comment added

File: engine/api/api.go (hunk 10 above)
Comment (yesnault) [reply engine/api/api.go]: it's now an array

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16387.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔

**Review by sguiheux — APPROVED** [review]

**Review by fsamin — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16389.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16389.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16389.2 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16389.3 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✘  
  * Standalone Tests ✔

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16392.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✘

**Comment by ovh-cds** [comment]
CDS Report build-all-cds#16392.1 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build OS-Ansible-Inventory ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test SDK Rust ✔ 
  * Test UI ✔ 
* Package
  * Docker ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✘

**Pull request merged**